### PR TITLE
feat(AggregateRouter): support native mongodb syntax in aggregation pipelines

### PR DESCRIPTION
**Pull request by RaschidJFR**
### New Pull Request Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Issue Description

Related issue: #7338 

> In the JS SDK Parse.Query.aggregate() works really well in runtime, but the fact that stage names differ from the built-in MongoDB’s by the $ symbol represents an issue for compatibility, for example, when importing/exporting/debugging a pipeline in other tools like MongoCompass.

### TODOs before merging

- [x] Add test cases
- [x] Add entry to changelog
- [x] Add changes to documentation (guides, repository pages, in-code descriptions)

### CODE REVIEW

**Comment by mtrezza** [comment]
Thanks for this PR!

To keep the code base simple and prevent potential issues in the future we should deprecate the current way of writing stage names without leading dollar sign.

We have a [Deprecator](https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md#breaking-changes) to phase-out features. We are now in a special situation before the 5.0.0 release where we are more graceful with allowing breaking changes without deprecation. If you feel like adding the deprecation to make it easier for developers to adjust to the change until 6.0.0, please feel free to add it, otherwise you may as well make it a sudden breaking change.

File: CHANGELOG.md
Comment (mtrezza) [inline CHANGELOG.md:105]: Depending on how you want to proceed, this either goes under the breaking changes section (without deprecation) or notable change (with deprecation).

File: src/Routers/AggregateRouter.js
Comment (mtrezza) [inline src/Routers/AggregateRouter.js:109]: Depending on how you want to proceed, remove feature of adding a dollar sign (without deprecation) or add the deprecation warning (with deprecation).

**Review by mtrezza — CHANGES_REQUESTED** [review]

File: CHANGELOG.md
Comment (RaschidJFR) [reply CHANGELOG.md:105]: I'll go for the deprecation warning. We all appreciate lowering the breaking change amount.

File: src/Routers/AggregateRouter.js
Comment (RaschidJFR) [reply src/Routers/AggregateRouter.js:109]: Thanks for the guidance. I had a look at the [Deprecator](https://github.com/parse-community/parse-server/blob/master/src/Deprecator/Deprecator.js#L35). But if I'm getting it correctly, I would need to add the list of currently available aggregation stages in latest MongoDB 4.x, right? But that is a dynamic list, so if more stages get added, they won't be caught as a deprecation warning.

In this case, I think it would be better to test just for the missing `$`. If needed I can make the changes in Deprecator for that but I doubt it may be used in another case.

Any thoughts?

Comment (mtrezza) [reply src/Routers/AggregateRouter.js:109]: > In this case, I think it would be better to test just for the missing $

Yes, makes sense.

> If needed I can make the changes in Deprecator for that but I doubt it may be used in another case.

I already expected that we'll need to add this to the deprecator, but I was waiting for the first case to implement this - and here it is. The concept of the deprecator is to define deprecations centrally, which is easy deprecating for Parse Server Options. For deprecations that are detected "in-code" I am still thinking what the best approach is. I would want to avoid spreading deprecation definitions all over the place.

**Comment by RaschidJFR** [comment]
Hey @mtrezza should I drop the backward compatibility and the deprecation warning in order to merge this or should we keep this open for a bit longer?

**Comment by mtrezza** [comment]
I have added the deprecation method required in https://github.com/parse-community/parse-server/pull/7451. Could you take a look to see if that is applicable to log a depr. warning for this PR?

**Comment by RaschidJFR** [comment]
Thanks @mtrezza . It works for this case. I'll just wait for the PR to be merged and apply it.

**Comment by mtrezza** [comment]
Could you please review the checkboxes in the PR template and check the ones that are done and delete the ones that don't apply?

**Comment by mtrezza** [comment]
@RaschidJFR The deprecation PR has been merged.

**Review by mtrezza — CHANGES_REQUESTED** [review]
See https://github.com/parse-community/parse-server/issues/7338#issuecomment-881184514

**Comment by RaschidJFR** [comment]
@mtrezza I modified the spec to test separately for `_id` and `$`. This should do. I think the rest of the test cases should eventually be updated to the new syntax (not now as it would compromise the coverage). Mark as to-do?

**Comment by mtrezza** [comment]
Amazing! Yes, you could add this as TODO here for the record. Eventually we'll have to do that once we remove the deprecation, so we have enough time to do that at some point.

**Comment by RaschidJFR** [comment]
What do you think @mtrezza ? Are we ready to merge?

**Comment by mtrezza** [comment]
Could you look at the failing test? It seems to be related to this change.

```
1) Parse.Query Aggregate testing invalid query group _id
  - Error: Timeout - Async function did not complete within 20000ms (set by jasmine.DEFAULT_TIMEOUT_INTERVAL)
```

**Comment by RaschidJFR** [comment]
Tests updated. For the record, I removed [this case](https://github.com/parse-community/parse-server/pull/7339/files#diff-ed82ab7ef5313a1884b38177f0a4bcc8fc7df356eee41d416eb15269971083a0L86), which doesn't seem to be required anymore, unless there is a specific reason for it and that I ignore.

Switching topic, I'm having troubles with a [failing case](https://github.com/parse-community/parse-server/runs/3287223873?check_suite_focus=true) in MongoDB 4.2. It does not fail when I run it locally, but it didn't pass on the CI. Is there something I need to look into?

**Comment by mtrezza** [comment]
> Tests updated. For the record, I removed this case, which doesn't seem to be required anymore, unless there is a specific reason for it and that I ignore.

makes sense

> Switching topic, I'm having troubles with a failing case in MongoDB 4.2. It does not fail when I run it locally, but it didn't pass on the CI. Is there something I need to look into?

The aggregation pipeline is a fairly new feature in MongoDB; it's possible that there was a change in that version.

**Comment by mtrezza** [comment]
Merged master, because the MongoDB versions have changed.

**Review by mtrezza — APPROVED** [review]
Thanks for this great PR! This should make aggregation pipeline development much easier now.

Before we merge this, I would like to create a deprecation list where we add this change as an entry. So we have a table where we can see this deprecation, when it was introduced and when the deprecated feature is planned to be removed. This should be for maintainers to know when to remove features, and for developers to better plan when changes become required to upgrade.

**Comment by mtrezza** [comment]
The PR is #7485, as soon as it's merged, you can just replace the `(none)` line with:

| Feature | Issue | Deprecation  | Planned Removal  | Status | Notes |
|---------|----|------------------|----------------------|----------|-------|
| Native syntax in MongoDB aggregation pipeline | [#7338](https://github.com/parse-community/parse-server/issues/7338) | 5.0.0 (2022) | 6.0.0 (2023) | deprecated | - |

**Review by mtrezza — APPROVED** [review]
Looks good to me!

**Comment by RaschidJFR** [comment]
@mtrezza thanks for fixing the typo, it slipped away 😅

**Comment by mtrezza** [comment]
Sure! This is ready for merge in my view, but I would like to get a second opinion, because it is quite a substantial breaking change. @sadortun what do you think?

**Comment by sadortun** [comment]
@mtrezza I'm not that familiar with pipelines but here's a few unrelated  thoughts.

1) Docs should indicate more clearly that pipelines are actually Mongo syntax. `pipelines use the same syntax as Mongo x.y, you can find more examples here`

2) it could be useful to have a new CLP to allow non-masterKey clients to use aggregates. Or a limited version of aggregates if it cause  a security issue.

**Comment by mtrezza** [comment]
Thanks @sadortun, apologies, I did not ask you first, whether you use MongoDB or are familiar with aggregation.

It seems what you are saying is that it could be better clarified that Aggregation Pipeline is a MongoDB specific term and feature. That would be a general issue, something we can keep in mind when improving the docs.

Regarding security, aggregation is just another query type that obeys the same permissions as any other query type.

So I take we are good to go here.

**Comment by sadortun** [comment]
> Regarding security, aggregation is just another query type that obeys the same permissions as any other query type.

Just to clarify, Docs mention `MasterKey is Required.` for `Query.aggregate` and `Query.distinct`, it would be nice to add 2 extra CLP `aggregate` and `distinct` to allow specific collections to be aggregated by some users/roles. 

If this makes sense,LMK,  I'll open a separate issue.

**Comment by mtrezza** [comment]
Merged - thanks Rashid for this PR!

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.0.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/5.0.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.0.0](https://github.com/parse-community/parse-server/releases/tag/5.0.0)

**Pull request merged**
